community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/logical/idp/cartesianProductsOrValueJoins.scala

```diff
@@ -49,28 +49,56 @@ case object cartesianProductsOrValueJoins extends JoinDisconnectedQueryGraphComp
 
     assert(plans.size > 1, "Can't build cartesian product with less than two input plans")
 
-    // Map of newly found plans to originating LHS and RHS plans to join
-    val connectedPlans: Map[PlannedComponent, (PlannedComponent, PlannedComponent)] = {
-      val joins = produceJoinVariations(plans, qg)
+    /*
+    To connect disconnected query parts, we have a couple of different ways. First we check if there are any joins that
+    we could do. Joins are equal or better than cartesian products, so we always go for the joins when possible.
 
-      if (joins.nonEmpty)
-        joins
-      else
-        produceCartesianProducts(plans, qg)
+    Next we can do an exhaustive search for how to combine the remaining query parts together. In-between each step
+    we check if any joins have been made available. This makes for better plans, but is exponentially expensive.
```

Comment:
This comment kind of makes it sound like it's the in-between checking for joins that causes the exponential increase. Is it? I figure from the PR comment that the exponential increase comes from the `exhaustive search` mentioned in the first sentence of this paragraph.

I suggest making this a bit clearer:

Next we perform an exhaustive search for how to combine the remaining query parts together. In-between each step we check if any joins have been made available. This exhaustive search makes for better plans, but is exponentially expensive.